nservicebus/messaging/message-type-detection.md

```diff
@@ -0,0 +1,18 @@
+---
+title: Message type detection
+reviewed:
+component: Core
+---
+
+In order to invoke the correct message handlers for incoming messages, NServiceBus needs to be able to map the incoming transport message to a [message type](/nservicebus/messaging/messages-events-commands.md).
+
+The mapping rules are as follows:
+
+1. If the message contains the [`NServiceBus.EnclosedMessageType` header](/nservicebus/messaging/headers.md#serialization-headers-nservicebus-enclosedmessagetypes), the header value will be used to find the message type. The header value must at least contain the [FullName](https://docs.microsoft.com/en-us/dotnet/api/system.type.fullname) of the message type but can also contain the [AssemblyQualifiedName](https://docs.microsoft.com/en-us/dotnet/api/system.type.assemblyqualifiedname). NServiceBus uses the AssemblyQualifiedName when emitting messages.
+
+1. If the header is missing, serializers can optionally infer the message type based on the message payload.
+
+## Serializers that support message type inference
+
+* [Xml](/nservicebus/serialization/xml.md#inferring-message-type-from-root-node-name)
+* [Json.NET](/nservicebus/serialization/newtonsoft.md#inferring-message-type-from-type)
```

Comment:
The issue I have with this addition is _how_ type inference works is very different from one serializer to another. For XML, the native message itself will contain the type. E.g.:

```xml
<Place Order>
  <OrderId>1234</OrderId>
</Place Order>
```

There's nothing special that needs to be done.

Though with Newtonsoft.Json that's not the case. A native message could look like this.

```json
{
  "OrderId":"1234"
}
```

For type inference to actually work, the native message must contain _extra_ information and the message to look like

```json
{
  "$type": "Namespace.PlaceOrder"
  "OrderId": "1234"
}
```

Therefore I suggest to either put a warning or expand in this document that caution must be exercised. People don't read the documentation thoroughly. They'll see "JSON" and "type inference" and will run with that.

Reply:
I've tried to clarify this now